### PR TITLE
[SPARK-7655][Core] Deserializing value should not hold the TaskSchedulerImpl lock

### PR DESCRIPTION
**Pull request by zsxwing**
We should not call `DirectTaskResult.value` when holding the `TaskSchedulerImpl` lock. It may cost dozens of seconds to deserialize a large object.

### CODE REVIEW

**Comment by zsxwing** [comment]
@rxin @yhuai @marmbrus

**Comment by rxin** [comment]
As discussed offline, let's update the pull request with more comments on where the lock is. That is not obvious just looking at the diff.

**Comment by rxin** [comment]
lgtm

**Comment by shivaram** [comment]
@rxin did you benchmark what / how much this improves things ? I've been meaning to debug these really slow `collect` operations and am wondering if this was the reason

**Comment by rxin** [comment]
I didn't do anything here.

**Comment by shivaram** [comment]
You reviewed the PR :). Anyways @zsxwing -- Did you benchmark any improvements due to this change with some workload ?

**Comment by zsxwing** [comment]
> You reviewed the PR :). Anyways @zsxwing -- Did you benchmark any improvements due to this change with some workload ?

Depends on your case. If you have only one job, or the values to be collected are very small, I think this PR improves nothing. But if you have multiple jobs. Deserializing multiple values is in parallel now. It may improve the performance.

File: core/src/main/scala/org/apache/spark/scheduler/TaskResult.scala [hunk 1]
```diff
@@ -40,6 +40,9 @@ class DirectTaskResult[T](var valueBytes: ByteBuffer, var accumUpdates: Map[Long
     var metrics: TaskMetrics)
   extends TaskResult[T] with Externalizable {
 
+  private var valueObjectDeserialized = false
```
Comment (markhamstra) [inline core/src/main/scala/org/apache/spark/scheduler/TaskResult.scala]: Shouldn't this be `@volatile`?

Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/scheduler/TaskResult.scala]: No. The two places calling `value()` are in the same thread.

Comment (markhamstra) [inline core/src/main/scala/org/apache/spark/scheduler/TaskResult.scala]: Ah, got it.  Yes, you're correct.  Thanks.

**Pull request closed without merging**